client/lib/interval/index.js

```diff
@@ -0,0 +1,64 @@
+import React, { PropTypes } from 'react';
+
+/**
+ * Calls a given function on a given interval
+ */
+export default React.createClass( {
+	displayName: 'Interval',
+
+	propTypes: {
+		onTick: PropTypes.func.isRequired,
+		milliseconds: PropTypes.number,
+		leading: PropTypes.bool,
+		pauseWhenHidden: PropTypes.bool,
+		children: PropTypes.element
+	},
+
+	getDefaultProps: () => ( {
+		milliseconds: 30000,
+		leading: true,
+		pauseWhenHidden: true
+	} ),
+
+	getInitialState: () => ( {
+		timer: null
+	} ),
+
+	componentDidMount() {
+		this.start()
+	},
+
+	componentWillUnmount() {
+		this.stop()
+	},
+
+	componentDidUpdate() {
+		this.start();
+	},
+
+	run() {
+		clearTimeout( this.state.timer );
+
+		if ( document.hidden && this.props.pauseWhenHidden ) {
+			return this.setState( { timer: null } );
+		}
+
+		this.setState( { timer: setTimeout( this.run, this.props.milliseconds ) } );
+
+		this.props.onTick();
```

Comment:
Thanks for the comments @omarjackman! This is probably unclear what exactly my goal is here. In this case the callback _shouldn't_ depend on the state and the call to `setState` is merely there so we can cancel the timeout if we need to. The call to `setState()` should queue up a JavaScript task that will be executed after the current call stack, meaning it should actually run _after_ `onTick()`, but as it is now the call to `setTimeout()` happens _before_ `onTick()`.

I suppose one way I could clear this up is to move the `setState()` call after `onTick()`, and probably the reason I didn't want to do that originally was to prevent against costly `onTick()` executions messing with the interval. _For example_, if the timeout were five seconds and `onTick()` took one second to complete, then we would end up with five seconds _between_ calls to `onTick()` instead of firing off `onTick()` once every five seconds.

The main point here is to both fire off the `onTick()` action and simultaneously schedule a new run some time in the future.